thunder/core/transform_common.py

```diff
@@ -363,3 +368,105 @@ class PostOptimizationTransform(Transform, ABC):
     @abstractmethod
     def transform_trace(self, computation_trace: Trace, **kwargs):
         pass
+
+
+def functionalize_inplace_ops(computation_trace: Trace) -> list[Trace]:
+    """Functionalize in-place ops in ``computation_trace``.
+
+    In thunder, an in-place is an out-of-place or functional op followed by :func:`~thunder.core.prims.copy_`.
+    This function replaces such in-place ops with out-of-place ops.
```

Comment:
"... only if the in-place argument is intermediate to the trace", right?

Reply:
Oh, I see later "functionalization is not applied, if any of an in-place op's arguments are ``computation_trace.args`` or ``computation_trace.kwargs``."

Reply:
it seems that BatchNorm's `num_batches_tracked` tensor update is expressed as `ltorch.add_(num_batches_tracked, 1)` and the tensor is an arg. so this makes sense to me. also, if one or more of args & kwargs are updated in an in-place manner, then I guess there's some intention so I'm not inclined to ban such cases